core/templates/dev/head/pages/community-dashboard-page/question-opportunities/question-opportunities.directive.ts

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for question opportunities.
+ */
+
+require(
+  'pages/community-dashboard-page/opportunities-list/' +
+  'opportunities-list.directive.ts');
+require(
+  'pages/community-dashboard-page/services/' +
+  'contribution-opportunities.service.ts');
+
+angular.module('oppia').directive('questionOpportunities', [
+  'UrlInterpolationService', 'MAX_QUESTIONS_PER_SKILL',
+  function(UrlInterpolationService, MAX_QUESTIONS_PER_SKILL) {
+    return {
+      restrict: 'E',
+      scope: {},
+      bindToController: {},
+      templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+        '/pages/community-dashboard-page/question-opportunities/' +
+      'question-opportunities.directive.html'),
+      controllerAs: '$ctrl',
+      controller: [
+        'ContributionOpportunitiesService',
+        function(ContributionOpportunitiesService) {
+          var ctrl = this;
+          ctrl.opportunities = [];
+          ctrl.opportunitiesAreLoading = true;
+          ctrl.moreOpportunitiesAvailable = true;
+          var updateWithNewOpportunities = function(opportunities, more) {
+            for (var index in opportunities) {
+              var opportunity = opportunities[index];
+              var heading = opportunity.topic_name;
```

Comment:
Shouldn't we at some point be using ObjectFactories to convert the backend dicts to objects? Should we file an issue for this if we plan to do this later?

Reply:
I've created an issue at #7994. @nithusha21 should we create the domain object (From the dict) in the service or the directive? (In issue #7994 I've mentioned to do the conversion in the service file just after fetching the opportunities but I can change it as per your suggestion.)

Reply:
I am not very sure(about frontend stuff in general :P). But I think the objectFactory is the equivalent of the domain object and we expect to have an instance of the object factory instead of the dicts in the frontend? I don't know so please correct me if I am wrong...

Reply:
> But I think the objectFactory is the equivalent of the domain object and we expect to have an instance of the object factory instead of the dicts in the frontend? I don't know so please correct me if I am wrong...

Sorry for the misunderstanding, we are going to have a separate class/file for the object factory but I was asking whether we should convert the dict into domain object in the service layer or in the directive? In the backend, we do it in the service layer after fetching the models, so I just wanted to make sure we are going to do the same here, right?

Reply:
The convention is: the dict should be changed into a domain object as soon as possible after it's retrieved from the backend, before it is actually used anywhere. This is something that should happen in this PR rather than being left for later.

Reply:
I see most of the services are made just to fech the data and then we use the dict in the dicrective then we call other services which then converts those data into domain object. (For example: `ExplorationFeaturesBackendApiService.fetchExplorationFeatures` fetchs exploration data which is then used in the exploration controller in dict which then get passes to `ExplorationStatesService` and there we covert the dict into the domain object [I see this pattern in the many places]). So I wanted to make sure whether we want to do the dict to object conversion in the service layer? (Also, will it be fine for the service method name `fetchSomthing` to return domain object.)?

Reply:
Hm, looking at the usages of that function -- I do think it's against convention. There are some parts of the code where it isn't even converted into  a domain object.

I agree that we need a standard place for doing the dict-to-object conversion. I think you are correct that it's best to do it in the fetch* functions since then other services can use the return values from them without thinking about it. Would it be possible to fix the ones introduced in this PR accordingly here, and file an issue enumerating the other instances so that they could be fixed too?

Reply:
> There are some parts of the code where it isn't even converted into a domain object.

I agree, the same method `ExplorationFeaturesBackendApiService.fetchExplorationFeatures` is used in multiple places and somewhere we convert it to object and somewhere don't so I think the conversion just after fetching the data in the srevice will make sure that everywhere it's the same i.e, domain object not the dict.

>  Would it be possible to fix the ones introduced in this PR accordingly here

I'll make PR.

>  file an issue enumerating the other instances so that they could be fixed too?

I think this would need work on finding the places where we have to do this and create an issue that has sub-issues within them. Maybe we can follow this pattern while angular migration or maybe the angular-migration team should create an issue for the same so that it's fixed (on the go or after the migration)?

Reply:
> Maybe we can follow this pattern while angular migration or maybe the angular-migration team should create an issue for the same so that it's fixed (on the go or after the migration)

No, this is a different thing from the Angular migration and should be handled separately. I think the Angular migration team has enough stuff on their radar at the moment!

Reply:
Currently, this PR is blocking multiple other PRs which are in review, to unblock all other PRs I'm going to merge this PR and I'll create a separate PR for the domain objects. (I've assigned myself to #7994)

/cc @sagangwee

Reply:
Fixed this in https://github.com/oppia/oppia/pull/8063. [**Merged**]